### PR TITLE
Rename Contact

### PR DESCRIPTION
**Pull request by alexbarnsley**
Ability to rename a contact already in the account. Moved some stuff around so I didn't create any smelly code.

### CODE REVIEW

**Comment by alexbarnsley** [comment]
@fix - I won't accept any rejections :trollface:

File: client/app/src/components/addressbook/addressbook.controller.js [hunk 1]
```diff
@@ -63,6 +63,55 @@
...
+
+      var newContact = { name: name, address: address };
+      if (!isRename && self.addressExists(address)) {
+        self.showToast('A Contact with this Address already exists', address, true);
+        return;
+      }
+      if (self.contactExists(name)) {
+        self.showToast('A Contact with this Name already exists', name, true);
+        return;
+      }
```
Comment (spresnal) [inline client/app/src/components/addressbook/addressbook.controller.js]: Would be nice to pull all this error handling out into a method. Then follow the error handling with a single return if possible.

Might consider adding some unit tests for the logic as well to make sure everything is being handled appropriately.

File: client/app/src/components/addressbook/addressbook.controller.js [hunk 2]
```diff
@@ -63,6 +63,55 @@
...
+      self.save();
+      if (!suppressNotice) {
+        self.showToast('Contact successfully added', name, false);
+      }
+      if (typeof callback === 'function') {
+        callback();
+      }
+
+      return true;
+    }
```
Comment (spresnal) [inline client/app/src/components/addressbook/addressbook.controller.js]: Can this method be broken into smaller unit-testable chunks? Do we need to return false when an error comes up above?

File: client/app/src/components/addressbook/addressbook.controller.js [hunk 3]
```diff
@@ -157,25 +169,40 @@
...
+              self.contacts[i].address = address;
+              contactSaved = true;
+            }
+          }
+          if (contactSaved) {
+            self.showToast('Contact successfully saved', name, false);
+          } else {
+            self.showToast('Could not find the Contact to save', name, true);
+          }
+        } else {
```
Comment (spresnal) [inline client/app/src/components/addressbook/addressbook.controller.js]: A lot of possible code paths here. Unit tests? 😈

**Comment by alexbarnsley** [comment]
Thanks @spresnal - i'll implement some tests a bit later!

**Comment by alexbarnsley** [comment]
@spresnal what are your thoughts on doing these tests? I'll basically need to have data in the "client" such as wallets and contacts already before performing my tests. Would I set this up at the start of each test? I don't really write many unit tests 😬

**Comment by spresnal** [comment]
I wanted to put some example tests together, but ran out of time. Anyways, I'll try to answer your question:

>I'll basically need to have data in the "client" such as wallets and contacts already before performing my tests. Would I set this up at the start of each test?

The solution to this is stubbing out those method calls. We want to test code specifically within addressbook.controller.js and not the logic of any of the services being injected through the angular constructor.

You can either set that up a single time for all the tests (before) or before/after each (beforeEach/afterEach). It depends on the cost of the operation (in this case inexpensive) and whether the tests change the state of the object. For example, if a few tests are adding contacts to storage this may cause other unit tests to fail (in this case you'd want to reset the state of your storageService mock before/after each test). On the other hand, if you're only reading from storageService for all these tests you'd want to just use a before.

I would look into sinonjs (you may have to add it to the project) for stubbing out methods. Here's an example of stubbing an API call https://github.com/spresnal/league-of-legends-seed/blob/Basic-Setup/server/api/leagueApis/championMasteryApi.test.js. You can find a lot of other (better) examples with some searching.

Hope this helps!

**Comment by alexbarnsley** [comment]
Thanks for the explanation @spresnal (I forgot to say). I've been trying to get the first service test working. It's close but whenever I instantiate accountService (or networkService), it does a `var ark = require('arkjs');` which I've got working, but that then does `require('./lib/transactions/crypto.js')` which fails:

```
npm test


> ArkClient@1.4.1 test C:\Users\alexj\projects\ark-desktop
> karma start test/karma.conf.js --single-run

22 10 2017 14:34:24.594:INFO [karma]: Karma v1.7.1 server started at http://0.0.0.0:9876/
22 10 2017 14:34:24.597:INFO [launcher]: Launching browser PhantomJS with unlimited concurrency
22 10 2017 14:34:24.603:INFO [launcher]: Starting browser PhantomJS
22 10 2017 14:34:27.565:INFO [PhantomJS 2.1.1 (Windows 8 0.0.0)]: Connected on socket 0pHKJ1v9dNdoibm4AAAA with id 88950999
PhantomJS 2.1.1 (Windows 8 0.0.0) ERROR
  Error: Cannot find module './lib/transactions/crypto.js'
  at C:/Users/alexj/projects/ark-desktop/node_modules/mocha/mocha.js:1


PhantomJS 2.1.1 (Windows 8 0.0.0): Executed 0 of 2 SUCCESS (0 secs / 0 secs)
npm ERR! Test failed.  See above for more details.
```

Tried doing absolute path and still nothing. faustbrian is helping me try and get it sorted so hopefully some progress soon!

**Comment by luciorubeens** [comment]
Hmm I think the contacts should be stored as a json instead of array, something like this:

`var contacts = { "address": "name", "address2": "name" }`

So to add/edit would be a little bit easier: `contacts[address] = name`
And to remove: `delete contacts[address]`

**Comment by spresnal** [comment]
@alexbarnsley sorry if I potentially created some merge conflicts for you. I didn't realize the changes here still hadn't gone in! Anyways, if you reference #379 , you may find what the issue is on your end.

**Comment by alexbarnsley** [comment]
Don't worry about conflicts! I can sort those. May take a look at your test PR as may be helpful to with the tests I needed here! Thanks!

**Comment by alexbarnsley** [comment]
@faustbrian been intentionally leaving as lowest on my priority list. Will close as a lot has changed and don't remember since was before I was part of the team. Will leave the branch in case anyone wants to pick it up, and will create a git issue for it.

**Pull request closed without merging**